File: codes/ruby/chapter06/answer_A42.rb
# 整数の組 (a, b) が決まったときの、参加可能な生徒数を返す関数
def get_score(a, b, a_arr, b_arr, k)
  cnt = 0
  a_arr.each_with_index do |value, index|
    if a <= value && value <= a + k && b <= b_arr[index] && b_arr[index] <= b + k
      cnt += 1
    end
  end
  return cnt
end

# 入力
n, k = gets.split.map(&:to_i)
a = []
b = []
n.times do
  a_val, b_val = gets.split.map(&:to_i)
  a << a_val
  b << b_val
end

# (a, b) の組を全探索
answer = 0
(1..100).each do |a_val|
  (1..100).each do |b_val|
    score = get_score(a_val, b_val, a, b, k)
    answer = [answer, score].max
  end
end

# 出力
puts answer

File: codes/go/chapter06/answer_A42.go
package main

import (
	"fmt"
)

func main() {
	var N, K int
	fmt.Scan(&N, &K)
	A := make([]int, N+1)
	B := make([]int, N+1)
	for i := 1; i <= N; i++ {
		fmt.Scan(&A[i], &B[i])
	}

	// (a, b) の組を全探索
	var Answer int
	for a := 1; a <= 100; a++ {
		for b := 1; b <= 100; b++ {
			Score := GetScore(a, b, N, K, A, B)
			if Score > Answer {
				Answer = Score
			}
		}
	}

	// 出力
	fmt.Println(Answer)
}

// 整数の組 (a, b) が決まったときの、参加可能な生徒数を返す関数
func GetScore(a int, b int, N int, K int, A []int, B []int) int {
	cnt := 0
	for i := 1; i <= N; i++ {
		if a <= A[i] && A[i] <= a+K && b <= B[i] && B[i] <= b+K {
			cnt += 1
		}
	}
	return cnt
}
